### PR TITLE
Added rule to prefer min()/max() over sorted().first and sorted().last

### PR DESCRIPTION
**Pull request by tomquist**
Implements #1932

### CODE REVIEW

**Comment by SDGGiesbrecht** [comment]
@tomquist,

> Use confusing date format which seems to be preferred by SwiftLint

Dass Amerikaner “confusing” sind weiß jeder. ☺ Doch in diesem Fall ist das deutsche Format genauso sinnlos. ☺ Die [ISO](https://de.wikipedia.org/wiki/ISO_8601#Zusammenfassung) weiß was sie tut. ☺

Aber am Ende liegt Realms Hauptsitz in San Fransisco – also machen wir alle trotzdem mit.

**Comment by jpsim** [comment]
Thanks for the PR! We really need to come up with some cleaner utilities to match multiple function calls. This PR contains duplicated code very similar to what's used to detect `.first != nil` and `.filter { }.first`.

Could you possibly refactor `FirstWhereRule`, `ContainsOverFirstNotNilRule` and `SortedFirstLastRule` to share common code?

**Comment by tomquist** [comment]
I gave it a try. Not sure whether this is the best approach though.

File: Source/SwiftLintFramework/Protocols/CallPairRule.swift [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+        let contents = file.contents.bridge()
+        let structure = file.structure
+
+        let violatingLocations: [Int] = firstRanges.flatMap { range in
+            guard let bodyByteRange = contents.NSRangeToByteRange(start: range.location,
+                                                                  length: range.length),
+                case let firstLocation = range.location + range.length - 1,
+                let firstByteRange = contents.NSRangeToByteRange(start: firstLocation,
+                                                                 length: 1) else {
+                                                                    return nil
```
Comment (jpsim) [inline Source/SwiftLintFramework/Protocols/CallPairRule.swift]: The general style used elsewhere in this project is to indent the body of `if`/`guard`/block statements by one indentation level more than the parent level. I know Xcode makes this difficult, but I'd like to stay consistent with the current style for now.

Same with the call to `methodCall` below.

File: Source/SwiftLintFramework/Protocols/CallPairRule.swift [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+//  Copyright © 2017 Realm. All rights reserved.
+//
+
+import SourceKittenFramework
+
+public protocol CallPairRule: Rule {}
+
+extension CallPairRule {
+
+    public func validate(file: File,
```
Comment (jpsim) [inline Source/SwiftLintFramework/Protocols/CallPairRule.swift]: It would be very useful to document this function.

**Comment by jpsim** [comment]
Looks like that code sharing fixed missing violations in the first-where rule! It's paying off already. Could you please add a new changelog entry in the bug fixes section covering that?

File: CHANGELOG.md [hunk 3]
```diff
@@ -10,6 +10,10 @@
 
 ##### Enhancements
 
+* Make `first_where` and `contains_over_first_not_nil` rules also detect cases
```
Comment (marcelofabri) [inline CHANGELOG.md]: I'd add this under bug fixes

**Comment by jpsim** [comment]
Thanks for the great PR!

**Comment by tomquist** [comment]
Thanks for guiding me!

**Pull request merged**